File: backend-go-fiber/src/models/transaction.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Transaction represents a money transfer transaction
type Transaction struct {
	ID          primitive.ObjectID `bson:"_id,omitempty" json:"_id,omitempty"`
	FromAccount string             `bson:"fromAccount" json:"fromAccount" validate:"required"`
	ToAccount   string             `bson:"toAccount" json:"toAccount" validate:"required"`
	Amount      float64            `bson:"amount" json:"amount" validate:"required,gt=0"`
	Description string             `bson:"description" json:"description"`
	Status      string             `bson:"status" json:"status" validate:"required,oneof=pending success failed"`
	InitiatedBy primitive.ObjectID `bson:"initiatedBy" json:"initiatedBy"`
	Date        time.Time          `bson:"date" json:"date"`
	CreatedAt   time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt   time.Time          `bson:"updatedAt" json:"updatedAt"`
}

File: backend_fastapi/app/models/transaction.py
"""
Transaction model for MongoDB.
Defines the Transaction schema and database operations.
"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from bson import ObjectId
from app.models.user import PyObjectId


class TransactionBase(BaseModel):
    """Base transaction schema."""
    fromAccount: str
    toAccount: str
    amount: float = Field(..., gt=0)
    description: str = ""
    status: str = Field(default="pending", pattern="^(pending|success|failed)$")


class TransactionCreate(BaseModel):
    """Schema for creating a new transaction."""
    toAccount: str
    amount: float = Field(..., gt=0)
    description: str = ""


class TransactionInDB(TransactionBase):
    """Transaction schema as stored in database."""
    id: Optional[PyObjectId] = Field(default=None, alias="_id")
    initiatedBy: Optional[PyObjectId] = None
    date: datetime = Field(default_factory=datetime.utcnow)
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}


class TransactionResponse(TransactionBase):
    """Transaction schema for API responses."""
    id: str = Field(alias="_id")
    initiatedBy: Optional[str] = None
    date: datetime
    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None

    class Config:
        populate_by_name = True
        json_encoders = {ObjectId: str}
